name: Python CI

on:
  push:
    branches: [ master ]
  pull_request:
    branches:
      - '**'
  workflow_dispatch:

concurrency:
  group: "${{ github.workflow }}-${{ github.ref }}"
  cancel-in-progress: true

jobs:
  tests:
    runs-on: ${{ matrix.os }}
    services:
      # Using SQLite3 for integration tests throws `django.db.utils.OperationalError: database table is locked: workbench_xblockstate`.
      mysql:
        image: mysql:8
        env:
          MYSQL_ROOT_PASSWORD: rootpw
        ports:
          - 3307:3306
    strategy:
      fail-fast: false
      matrix:
        os: [ubuntu-20.04]
        toxenv: [py38-django32, py38-django42, py311-django42, integration32, integration42, quality]

    steps:
    - name: checkout repo
      uses: actions/checkout@v4

    - name: setup python
      uses: actions/setup-python@v5
      with:
        python-version: |
          3.8
          3.11

    - name: Install Required System Packages
      if: ${{ startsWith(matrix.toxenv, 'integration') }}
      run: |
        sudo apt-get update
        sudo apt-get install -y xvfb

    - name: Install Dependencies
      run: make requirements

    - name: Run Tests
      env:
        TOXENV: ${{ matrix.toxenv }}
      run: tox

    - name: Run coverage
      if: matrix.python-version == '3.8' && matrix.toxenv == 'py38-django32'
      uses: codecov/codecov-action@v4
      with:
        token: ${{ secrets.CODECOV_TOKEN }}
        flags: unittests
        fail_ci_if_error: true
